### PR TITLE
feat(testing): add toMatch and toBeCloseTo matchers (#1675, #1676)

### PR DESCRIPTION
**Pull request by t0k0sh1**
## Summary

Adds two Jest-style test matchers for the v0.0.22 milestone:

- **`toMatch(pattern: str)`** (#1676) — regex string matcher backed by
  `__ry_regex_is_match` (unanchored; use `^` / `$` to anchor). Both
  `actual` and `pattern` must be `str`. Invalid patterns surface as
  runtime panics with the regex engine's error message.
- **`toBeCloseTo(value)`** / **`toBeCloseTo(value, decimals)`** (#1675) —
  float approximate equality matcher: `|actual - value| < 0.5 * 10^-decimals`.
  Makes `expect(0.1 + 0.2).toBeCloseTo(0.3)` pass even though strict
  `toEq` (`FCmpOEQ`) does not. `decimals` defaults to `2` and must be a
  non-negative integer literal in `[0, 15]` (the upper bound is the f64
  mantissa precision; larger values would underflow the threshold to
  `0` and silently degrade to strict equality). Both operands accept
  `int` or `float`; mixed types are promoted to `f64`.

The two issues were originally planned as separate PRs but combined
here at the user's request after the parser/codegen edits proved to
share the same dispatch sites.

## Implementation notes

- `ExpectStmt` gains a `std::vector<ExprPtr> extra_args` field so
  matchers with multiple arguments (currently only `toBeCloseTo`'s
  optional `decimals`) need not overload `expected`.
- Parser dispatches `toBeCloseTo` as a special-case before the
  `matchers_with_arg` whitelist (since it accepts an optional second
  argument); `toMatch` is added to the existing `matchers_with_arg` set.
- Codegen rejects non-conforming operands at compile time:
  - `toMatch`: non-`str` actual or pattern
  - `toBeCloseTo`: non-numeric operand, non-literal `decimals`,
    or `decimals` outside `[0, 15]`
- Negative `decimals` (e.g. `-1`) arrives as a `UnaryExpr`-wrapped
  `NumberExpr`, naturally failing the `std::get_if<NumberExpr>` check
  and producing the "must be a plain integer literal" diagnostic.
- Updated `.claude/rules/tests-rejection-tdd.md` to reflect the new
  matcher count (19 → 20) and dispatch chain end (`toMatch` →
  `toBeCloseTo`).

## Test plan

- [x] 18 spec tests in `tests/spec/matchers.test.ry` (toMatch 8 + toBeCloseTo 10)
- [x] 4 happy-path C++ tests in `tests/test_codegen_advanced.cpp`
- [x] 7 rejection tests in `tests/test_codegen_fail.cpp` (toMatch 2 + toBeCloseTo 5)
  — direct triggers per `.claude/rules/tests-rejection-tdd.md` P8
- [x] `./build/ry_tests` — 2153 tests pass
- [x] `./build/ry test -p` — 178 test files, 0 failures
- [x] ASan + UBSan (`./build-asan/ry_tests` + `./build-asan/ry test -p`) — pass
- [x] TSan C++ (`./build-tsan/ry_tests`) — pass; Ry self-test warn-only per policy
- [x] clang-tidy — exit 0
- [x] cppcheck — exit 0
- [x] libFuzzer `fuzz_parser` — 60s, 43797 runs, no crashes
- [x] Manual: invalid regex (`expect("x").toMatch("(")`) panics with
      `regex error: unmatched '(' in pattern '('`
- [x] Manual: `toBeCloseTo` boundary (`|1.0 - 1.006| > 0.005`) fails
      with `expected close to 1.006 (decimals=2), got 1.0`

Closes #1675
Closes #1676

## Summary by CodeRabbit

* **New Features**
  * Added toBeCloseTo(value[, decimals]) for approximate numeric comparisons (decimals 0–15).
  * Added toMatch(pattern) for regex-based string matching.

* **Documentation**
  * Updated testing reference and changelog entries documenting both matchers and decimals rules.

* **Tests**
  * Added comprehensive tests covering behavior, edge cases, and validation.

* **Other**
  * Improved failure messages to show expected strings (includes decimals where applicable).

[Review Change Stack](https://app.coderabbit.ai/change-stack/t0k0sh1/ry/pull/1691)

### CODE REVIEW

File: src/codegen_test.cpp [hunk 1]
```diff
@@ -835,6 +840,72 @@ void CodeGen::emitStmt(ExpectStmt &s) {
...
+            if (!n || !n->suffix.empty())
+                codegenError("line " + std::to_string(s.loc.line) +
+                    ": toBeCloseTo: 'decimals' must be a plain integer literal");
+            // NumberExpr.value is the unsigned bit pattern of a non-negative magnitude;
+            // negative literals arrive as UnaryExpr-wrapped NumberExpr and fail the
+            // get_if check above. So `n->value >= 0` is implied here.
+            if (n->value > 15)
+                codegenError("line " + std::to_string(s.loc.line) +
+                    ": toBeCloseTo: 'decimals' must be in [0, 15]");
+            decimals = static_cast<int64_t>(n->value);
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply src/codegen_test.cpp]: Fixed (will be in the next commit). Switched to a `uint64_t decimalsMag` comparison in `src/codegen_test.cpp` so that oversized literals (e.g. `UINT64_MAX`, which `tryParseIntLiteral` stores as `int64_t(-1)` per the parser convention) no longer wrap past the signed range and bypass the bound. Negative literals still arrive as `UnaryExpr`-wrapped `NumberExpr` and fail the `get_if<NumberExpr>` check above, so the magnitude here is genuinely non-negative. Verified by `ExpectToBeCloseToRejectsDecimalsOverflow` (UINT64_MAX) and `ExpectToBeCloseToRejectsNegativeDecimals` (`-1`) — both pass.

File: changelog.d/1675-tobecloseto-matcher.md [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+- `expect(actual).toBeCloseTo(expected)` /
+  `expect(actual).toBeCloseTo(expected, decimals)` — new test matcher
+  for float approximate equality, modeled on Jest's `toBeCloseTo`.
+  Asserts `|actual - expected| < 0.5 * 10^-decimals`, which makes
+  `expect(0.1 + 0.2).toBeCloseTo(0.3)` pass even though strict
+  `toEq` (FCmpOEQ) does not. `decimals` defaults to `2` and must be
+  a non-negative integer literal in `[0, 15]` (the upper bound is
+  the f64 mantissa precision; larger values would underflow the
+  threshold to `0` and silently degrade to strict equality). Both
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply changelog.d/1675-tobecloseto-matcher.md]: Fixed (will be in the next commit). Reworded `changelog.d/1675-tobecloseto-matcher.md` to drop the incorrect "underflow to 0" claim. The new wording: "the upper bound is tied to practical `f64` precision; larger values no longer provide meaningful decimal-place guarantees because adjacent representable doubles differ by more than `0.5 * 10^-decimals`." That matches the actual reason for the cap (precision exhaustion at ~15-17 significant digits, not arithmetic underflow — `0.5 * 10^-16 ≈ 5e-17` is well within the f64 normal range).
